File: ios/SecretaryVaultAccess/Sources/SecretaryVaultAccess/SyncBadgeState.swift
import Foundation

/// What the sync badge shows. Pure presentation state derived from the VM's
/// inputs; carries no secrets.
public enum SyncBadgeState: Equatable, Sendable {
    case neverSynced
    case synced(sinceMs: UInt64)   // from SyncStatus.lastStateWriteMs
    case changesDetected           // monitor raised pendingChanges
    case reviewNeeded              // a prior pass returned conflictsPending
    case syncing
}

/// Derive the badge state. Precedence (first match wins): in-progress →
/// conflict awaiting review → advisory change detected → last-synced → never.
public func syncBadgeState(
    inProgress: Bool,
    pendingChanges: Bool,
    hasPendingConflict: Bool,
    status: SyncStatus?
) -> SyncBadgeState {
    if inProgress { return .syncing }
    if hasPendingConflict { return .reviewNeeded }
    if pendingChanges { return .changesDetected }
    if let ms = status?.lastStateWriteMs { return .synced(sinceMs: ms) }
    return .neverSynced
}

File: android/vault-access/src/main/kotlin/org/secretary/sync/SyncBadgeState.kt
package org.secretary.sync

/**
 * The advisory sync state shown on the badge. Mirror of the iOS SyncBadgeState
 * (docs/superpowers/specs/2026-06-15-c3-ios-sync-ui-design.md). Rendering — including the
 * relative "synced N min ago" label derived from [Synced.sinceMs] — is a slice-5 UI concern;
 * this type only carries the discrete state.
 */
sealed interface SyncBadgeState {
    /** No sync state has ever been written for this vault. */
    data object NeverSynced : SyncBadgeState

    /** Last successful state write at [sinceMs] (epoch millis), for a relative-time label. */
    data class Synced(val sinceMs: ULong) : SyncBadgeState

    /** The change monitor raised a debounced "remote changes detected" signal. */
    data object ChangesDetected : SyncBadgeState

    /** A prior pass surfaced a tombstone dispute awaiting the user's decision. */
    data object ReviewNeeded : SyncBadgeState

    /** A pass is currently running. */
    data object Syncing : SyncBadgeState
}

/**
 * Pure derivation of the badge from the model's flags + the latest status snapshot.
 * Precedence (highest first): syncing → review → changes → synced → never. The single
 * [reviewNeeded] input collapses the two ways a review can be pending (the model supplies
 * `reviewNeededFlag || pendingConflict != null`): the sync-at-unlock path raises the flag
 * with no stashed conflict (password dropped), while the interactive path stashes one.
 */
fun syncBadgeState(
    inProgress: Boolean,
    pendingChanges: Boolean,
    reviewNeeded: Boolean,
    status: SyncStatus?,
): SyncBadgeState {
    val sinceMs = status?.lastStateWriteMs
    return when {
        inProgress -> SyncBadgeState.Syncing
        reviewNeeded -> SyncBadgeState.ReviewNeeded
        pendingChanges -> SyncBadgeState.ChangesDetected
        sinceMs != null -> SyncBadgeState.Synced(sinceMs)
        else -> SyncBadgeState.NeverSynced
    }
}
